add copy to clipboard icon to tooltips

Gkrumbach07 opened the pull request:
<!--- If this is a non-code change, this template is not required; reference any issues or top-level descriptions as needed -->
<!--- All code change PRs should relate to an issue, reference it here; see example below -->
<!--- Closes: #123 -->
closes: https://github.com/opendatahub-io/odh-dashboard/issues/1476

## Description
<!--- Describe your changes in detail; the what, the why, any findings, etc -->
<!--- Include any screenshots of changed UI; Include any gifs if it was a flow / UX change -->
added a copy icon to the resource name tooltip
_design:_
    - Add the helper text to the bottom of the tooltip
    - copy icon uses <Icon /> still to preserve padding, however this means i had to set the style of the pointer manually
    - there is currently no confirmation when copied
    - the BYONImages from the backend is not returned as a standard k8 object... so i added a eitherNotBoth to allow for manual input of name and kind or a k8 resource.

_Outdated_
<img width="386" alt="Screenshot 2023-07-20 at 2 23 12 PM" src="https://github.com/opendatahub-io/odh-dashboard/assets/12587674/0c6a78c8-4dfe-44a3-aa50-45287294b85b">

_Outdated_

https://github.com/opendatahub-io/odh-dashboard/assets/12587674/5e896138-bd9b-4395-9bdc-5eaa42ad637b


## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- go to notebook images in settings
- make sure it is not empty
- see help icon
- hover on help icon and click
- make sure it copied

## Test Impact
<!--- What tests have you done to covert implemented functionality -->
<!--- If tests are not applicable, explain why here -->
no impact - all labels are preserved

## Request review criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Commits have been squashed into descriptive, self-contained units of work (e.g. 'WIP' and 'Implements feedback' style messages have been removed)
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit tests & storybook for related changes)

If you have UI changes: 
<!--- You can ignore these if you are doing manifest, backend, internal logic, etc changes; aka non-UI / visual changes -->
- [x] Included any necessary screenshots or gifs if it was a UI change.
- [x] Included tags to the UX team if it was a UI/UX change.


@kaedward does this text look good on this?
@vconzola @kywalker-rh Is this what you were envisioning?

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

andrewballantyne commented on frontend/src/components/ResourceNameTooltip.tsx:
```diff
@@ -9,52 +9,87 @@ import {
+    navigator.clipboard.writeText(text);
+  };
+
+  const name = resourceName ?? resource.metadata?.name;
+  const kind = resourceKind ?? resource.kind;
```
```suggestion
  const name = resourceName ?? resource?.metadata.name;
  const kind = resourceKind ?? resource?.kind;
```
`resource` can be undefined -- not sure `metadata` ever will be in our use-case

Gkrumbach07 replied:
this is resolved by removing the option to manually give resourceName and resourceKind

andrewballantyne commented on frontend/src/components/ResourceNameTooltip.tsx:
```diff
@@ -9,52 +9,87 @@ import {
+  const [isHovered, setIsHovered] = React.useState(false);
+
+  const copyToClipboard = async (text: string) => {
+    navigator.clipboard.writeText(text);
```
Gosh something about this native functionality recalls a potential to throw an error if it fails -- perhaps we look into the docs behind it? If you see nothing, ignore this comment.

Gkrumbach07 replied:
there a few ways to copy, this is what we use elsewhere in the dashboard.

_Docs_
> The promise is rejected if the caller does not have permission to write to the clipboard.

So it returns a promise, which we dont handle, but in the case it does fail, it just wont copy.

Gkrumbach07 replied:
no longer relevant

andrewballantyne commented on frontend/src/components/ResourceNameTooltip.tsx:
```diff
@@ -9,52 +9,87 @@ import {
-type ResourceNameTooltipProps = {
-  resource: K8sResourceCommon;
-  children: React.ReactNode;
+type ResourceNameTooltipProps = EitherNotBoth<
+  { resource: K8sResourceCommon },
+  { resourceName: string; resourceKind: string }
+> & {
+  children?: React.ReactNode;
 };
```
Not sure we want to change the API 🤔 I wonder if you can just craft a "fake" object... we just need a metadata object really

andrewballantyne commented on frontend/src/components/ResourceNameTooltip.tsx:
```diff
@@ -9,52 +9,87 @@ import {
+              </Stack>
+            }
+          >
+            <Icon
+              onMouseEnter={() => setIsHovered(true)}
+              onMouseLeave={() => setIsHovered(false)}
+              onClick={() => {
+                if (name) {
+                  copyToClipboard(name);
+                }
+              }}
+              isInline
+              style={{ cursor: 'pointer' }}
+              aria-label="More info"
+              tabIndex={0}
+            >
+              {isHovered ? <CopyIcon /> : <OutlinedQuestionCircleIcon />}
+            </Icon>
```
I think as @DaoDaoNoCode mentioned on slack... you guys will have to work out this, [conflicts](https://github.com/opendatahub-io/odh-dashboard/pull/1529/commits/eaf0e2201672b465b8b0324bccdd693aec5fa247) be afoot... are the actions needed on the `Icon`?

Gkrumbach07 replied:
we ended up changing the design in a call with @vconzola and @DaoDaoNoCode in order to better align with both PRs and a UX aligned with Patternfly design guidelines
<img width="384" alt="Screenshot 2023-07-21 at 7 55 05 AM" src="https://github.com/opendatahub-io/odh-dashboard/assets/12587674/5ed13fb9-66ce-46e5-bbd5-d3ea6ca93eba">

https://github.com/opendatahub-io/odh-dashboard/assets/12587674/fa7c26e2-9e69-4bdc-a082-f8d5415e91b5

Gkrumbach07 replied:
ok so this is as a popover
<img width="441" alt="Screenshot 2023-07-21 at 8 29 51 AM" src="https://github.com/opendatahub-io/odh-dashboard/assets/12587674/e7df5961-0f42-4ac7-8070-0b77911c8ce6">

I can always revert back to the original idea

DaoDaoNoCode reviewed (approved):
lgtm

andrewballantyne commented on frontend/src/pages/BYONImages/BYONImagesTable.tsx:
```diff
@@ -330,6 +331,17 @@ export const BYONImagesTable: React.FC<BYONImagesTableProps> = ({ images, forceU
                       alignItems={{ default: 'alignItemsCenter' }}
                     >
                       <FlexItem>{image.name}</FlexItem>
+                      <FlexItem>
+                        <ResourceNameTooltip
+                          resource={{
+                            apiVersion: '',
+                            metadata: {
+                              name: image.id,
+                            },
+                            kind: 'ImageStream',
+                          }}
+                        />
+                      </FlexItem>
```
Why not wrap these two together? Seems odd to change the functionality of `ResourceNameTooltip` because you don't want to wrap the value?

Gkrumbach07 replied:
Oh i see use the children prop. Yeah, I went with this method to preserve the spacing in the flex box, bc there is an error that sometimes will show

andrewballantyne replied:
Keep the two FlexItems that predate you -- wrap the `image.name`

andrewballantyne commented on frontend/src/pages/BYONImages/BYONImagesTable.tsx:
```diff
@@ -330,6 +331,17 @@ export const BYONImagesTable: React.FC<BYONImagesTableProps> = ({ images, forceU
+                      <FlexItem>
+                        <ResourceNameTooltip
+                          resource={{
+                            apiVersion: '',
```
This looks very weird -- Can you not just use what we have in the model?

Gkrumbach07 replied:
What do you mean the model? It expects a k8 resource

andrewballantyne replied:
You're not adding version because it's not a real resource -- the model we use for the kind you're specifying has the data -- https://github.com/opendatahub-io/odh-dashboard/blob/main/frontend/src/api/models/openShift.ts#L17-L22

Gkrumbach07 replied:
So grab that from the constant to use? Even tho it doesn't get used

DaoDaoNoCode replied:
This is what I did on my branch
```
export const convertBYONImageToK8sResource = (image: BYONImage): K8sResourceCommon => ({
  kind: 'ImageStream',
  apiVersion: 'image.openshift.io/v1',
  metadata: {
    name: image.name,
    annotations: {
      'openshift.io/display-name': image.display_name,
    },
  },
});
```

andrewballantyne replied:
@Gkrumbach07 How is the next developer going to know you chose to put an empty string in a field that is required that we chose to use later in a generic component? If you use nothing, don't adjust the types to say it's optional, and just say "it's not used today" -- TS can't even help us later when we might choose to support the versions. You can hope we remember this, hope we check every reference (this is a highly used component, with this one edge case from legacy code), or just in general hope that when we use it later it supports being empty despite being required in the type.

We need to be mindful of the code we write "tomorrow" and not make our current code fail because we just chose to, otherwise we'll always be chasing bugs. Juntao's solution works fine and has some future proof mentality to it.

andrewballantyne commented on frontend/src/components/ResourceNameTooltip.tsx:
```diff
@@ -1,32 +1,35 @@
           removeFindDomNode
           position="right"
-          content={
+          showClose={false}
+          withFocusTrap={false}
```
Breaking a11y here. cc @DaoDaoNoCode @Gkrumbach07 

We want both of these.

Gkrumbach07 replied:
Hmm, this brings back that weird issue with the tooltip flashing

andrewballantyne replied:
I didn't see that in my testing -- what browser are you using?

Gkrumbach07 replied:
Chrome

andrewballantyne replied:
This is very odd -- this issue does not exist on PF docs. Can you log a bug for us to trace down the cause, I think we are misusing the component or we are out of date or there is some weird combo going on here.

andrewballantyne commented on frontend/src/components/ResourceNameTooltip.tsx:
```diff
@@ -1,32 +1,35 @@
 type ResourceNameTooltipProps = {
   resource: K8sResourceCommon;
-  children: React.ReactNode;
+  children?: React.ReactNode;
```
Undo

Gkrumbach07 commented:
@vconzola @kywalker-rh Here is an update of the current UI

This is the same popover you get from the BYON UI too.

A bit of a summary:
- We moved away from the UX of having the help icon switch to a copy icon as the curser covered the icon on hover, and there was no clear text we found that indicated that pressing the tooltip icon was how you copied the name.
- We then moved to adding the copy icon to the tooltip body, this was against PF guidelines (having user actions inside a tooltip)
- We moved to a popover to fix that issue, but we needed to then make the icon a button as popovers need to be pressed and not hovered. Icon buttons mess with the icon padding, so we made a custom `DashboardPopupIconButton` which removes this padding so it looks better inline with other icons
- We also then moved away from creating our own copy button inside the popover, and are now using PF inline compact copy field

<img width="499" alt="Screenshot 2023-07-24 at 8 40 29 AM" src="https://github.com/opendatahub-io/odh-dashboard/assets/12587674/897beabe-7faa-49a0-8838-097550468595">

https://github.com/opendatahub-io/odh-dashboard/assets/12587674/14193775-7861-4f24-9e2f-8bb37cce111b

DaoDaoNoCode reviewed (approved):
lgtm

vconzola commented:
This is definitely the correct solution from a UX and PF perspective. Thanks for doing this. I assume it will be applied to all places in the UI where we display resource names?

kywalker-rh commented:
Thanks for doing all this, and for all the back and forth, I know it was a lot. I think this is definitely the right solution.

andrewballantyne reviewed (approved):
Let us go ahead with this -- I think we need an issue (@Gkrumbach07 please do this) on our side to figure out why the popover snaps up to the top right before getting over the element... I imagine this is some overrides on styles, out of date dependency, or some mixture of components. We should investigate and log that with PF. It only happens on the first instance and then operates cleanly for the rest of the app.
